processor/metricstransformprocessor/metrics_transform_processor_helper.go

```diff
@@ -0,0 +1,375 @@
+			LabelValues:    keyToLabelValuesMap[key],
+			Points:         newPoints,
+		}
+		newTimeSeriesList[idxCounter] = newSingleTimeSeries
+		idxCounter++
+	}
+
+	return newTimeSeriesList
+}
+
+// analyzeTimeseriesForAggregation uses timestamp as an indicator to group aggregatable points in the group of timeseries, and determines the final value of the start timestamp for this groups of timeseries.
+// Returns a map that groups points by timestamps and the final start timestamp
+func (mtp *metricsTransformProcessor) analyzeTimeseriesForAggregation(timeseries []*metricspb.TimeSeries) (map[string][]*metricspb.Point, *timestamp.Timestamp) {
+	var startTimestamp *timestamp.Timestamp
+	// timestampToPoints maps from timestamp string to aggregatable points
+	timestampToPoints := make(map[string][]*metricspb.Point)
+	for _, ts := range timeseries {
+		if startTimestamp == nil || ts.StartTimestamp.Seconds < startTimestamp.Seconds || (ts.StartTimestamp.Seconds == startTimestamp.Seconds && ts.StartTimestamp.Nanos < startTimestamp.Nanos) {
+			startTimestamp = ts.StartTimestamp
+		}
+		for _, p := range ts.Points {
+			if points, ok := timestampToPoints[p.Timestamp.String()]; ok {
+				timestampToPoints[p.Timestamp.String()] = append(points, p)
+			} else {
+				timestampToPoints[p.Timestamp.String()] = []*metricspb.Point{p}
+			}
+		}
+	}
+	return timestampToPoints, startTimestamp
+}
+
+// aggregatePoints aggregates points in the groups provided in timestampToPoints by the specific caluculation indicated by aggrType and dataType
+// Returns a group of aggregated points
+func (mtp *metricsTransformProcessor) aggregatePoints(timestampToPoints map[string][]*metricspb.Point, aggrType AggregationType, dataType metricspb.MetricDescriptor_Type) []*metricspb.Point {
+	// initialize to size 0. unsure how many points will come out because distribution points are not guaranteed to be aggregatable. (mismatched bounds)
+	newPoints := make([]*metricspb.Point, 0)
+	for _, points := range timestampToPoints {
+		// use the timestamp from the first element because these all have the same timestamp as they are grouped by timestamp
+		timestamp := points[0].Timestamp
+		intPoint, doublePoint, distPoints := mtp.compute(points, aggrType, dataType)
+		if intPoint != nil {
+			newPoints = append(newPoints, &metricspb.Point{
+				Timestamp: timestamp,
+				Value:     intPoint,
+			})
+		} else if doublePoint != nil {
+			newPoints = append(newPoints, &metricspb.Point{
+				Timestamp: timestamp,
+				Value:     doublePoint,
+			})
+		} else {
+			for _, p := range distPoints {
+				newPoints = append(newPoints, &metricspb.Point{
+					Timestamp: timestamp,
+					Value:     p,
+				})
+			}
+		}
+	}
+	return newPoints
+}
+
+// compute attempts to merge points into one point based on the provided aggrType and dataType
+// Returns an int value, a double value and a slice of distribution values, and only one of these would contain values based on the dataType.
```

Comment:
Why not make this three functions then?

Reply:
I considered breaking this up by datatypes and aggrtypes, but I thought that might be a bit of over engineering? I personally really like small granular functions, so if you also think so, I would love to break this up.